pgjdbc/src/main/java/org/postgresql/jdbc/TypeInfoCache.java

```diff
@@ -177,267 +186,602 @@ public synchronized void addDataType(String type, Class<? extends PGobject> klas
-    }
-    String fullName = isArray ? pgTypeName.substring(0, pgTypeName.length() - 2) : pgTypeName;
-    String schema;
-    String name;
-    // simple use case
-    if (dotIndex == -1) {
-      schema = null;
-      name = fullName;
-    } else {
-      if (fullName.startsWith("\"")) {
-        if (fullName.endsWith("\"")) {
-          String[] parts = fullName.split("\"\\.\"");
-          schema = parts.length == 2 ? parts[0] + "\"" : null;
-          name = parts.length == 2 ? "\"" + parts[1] : parts[0];
-        } else {
-          int lastDotIndex = fullName.lastIndexOf('.');
-          name = fullName.substring(lastDotIndex + 1);
-          schema = fullName.substring(0, lastDotIndex);
-        }
-      } else {
-        schema = fullName.substring(0, dotIndex);
-        name = fullName.substring(dotIndex + 1);
-      }
-    }
-    if (schema != null && schema.startsWith("\"") && schema.endsWith("\"")) {
-      schema = schema.substring(1, schema.length() - 1);
-    } else if (schema != null) {
-      schema = schema.toLowerCase();
+    if (_getOidStatementComplexNonArray == null) {
+      String sql = "SELECT t.oid, n.nspname = ANY(current_schemas(true)), n.nspname,"
+          + " t.typinput = 'array_in'::regproc,"
+          + " t.typname, t.typelem, t.typtype, t.typdelim,"
+          + " e.typname, e.typtype, e.typdelim,"
+          + " arr.oid, e.typtype"
+          + "  FROM pg_catalog.pg_type t"
+          + "  JOIN pg_catalog.pg_namespace n ON t.typnamespace = n.oid"
+          + "  LEFT JOIN (SELECT s.r, (current_schemas(true))[r] AS nspname"
+          + "               FROM generate_series(1, array_upper(current_schemas(true), 1)) AS s (r)) AS sp"
+          + "    USING (nspname)"
+          + "  LEFT JOIN pg_catalog.pg_type arr ON (arr.typelem, arr.typinput) = (t.oid, 'array_in'::regproc)"
+          + "  LEFT JOIN pg_catalog.pg_type e ON t.typelem = e.oid"
+          + " WHERE t.typname = ? AND (n.nspname = ? OR ? IS NULL AND n.nspname = ANY (current_schemas(true)))"
+          + " ORDER BY sp.r LIMIT 1";
+      _getOidStatementComplexNonArray = _conn.prepareStatement(sql);
     }
-    if (name.startsWith("\"") && name.endsWith("\"")) {
-      name = name.substring(1, name.length() - 1);
-    } else {
-      name = name.toLowerCase();
-    }
-    oidStatementComplex.setString(1, name);
-    oidStatementComplex.setString(2, schema);
-    oidStatementComplex.setString(3, schema);
+    oidStatementComplex = _getOidStatementComplexNonArray;
+    oidStatementComplex.setString(1, typeName.typname());
+    oidStatementComplex.setString(2, typeName.nspname());
+    oidStatementComplex.setString(3, typeName.nspname());
     return oidStatementComplex;
   }
 
-  public synchronized int getPGType(String pgTypeName) throws SQLException {
-    Integer oid = _pgNameToOid.get(pgTypeName);
-    if (oid != null) {
-      return oid;
+  private PreparedStatement getArrayOidStatement(ParsedTypeName typeName) throws SQLException {
```

Comment:
does the read/write of mutable member variables in this method require synchronization, or is the assumption that this method is only called from other synchronized methods? If the latter, it might be helpful to document that assumption/requirement.

Reply:
Good catch. I'm adding comments.